### PR TITLE
Logcontext docs

### PR DESCRIPTION
**Pull request by richvdh**
A bunch of RST about how logcontexts work

### CODE REVIEW

File: docs/log_contexts.rst [hunk 1]
```diff
@@ -1,10 +1,441 @@
...
+Was all this really necessary?
+------------------------------
+
+The conventions used work fine for a linear flow where everything happens in
+series via ``defer.inlineCallbacks`` and ``yield``, but are certainly tricky to
+follow for any more exotic flows. It's hard not to wonder if we could have done
+something else.
+
+We're not going to rewrite Synapse now, so the following is entirely of
+acadamic interest, but I'd like to record some thoughts on an alternative
```
Comment (ara4n) [inline docs/log_contexts.rst]: academic

Comment (richvdh) [inline docs/log_contexts.rst]: I find the keys to the innermost secrets of the universe and all I get is a correction to a typo? Sheesh. Tough crowd. 

:)

Comment (ara4n) [inline docs/log_contexts.rst]: did you not see all my emoji reactions?!? ;p

**Comment by ara4n** [comment]
i would love to know from @glyph and the twisted guys whether we are missing a trick with Twisted and whether this is the Right Way or not to solve this...

**Comment by erikjohnston** [comment]
I believe glyph(?) helpfully pointed us at a library that does this in a better/less hacky way already, we just haven't had a chance to investigate it.

**Comment by ara4n** [comment]
yup, @markrwilliams pointed us at https://bitbucket.org/hipchat/txlocal from the nice folks at hipchat. Just wondering if there are other options for doing this...

**Comment by ara4n** [comment]
@glyph has left words of wisdom at https://twitter.com/glyph/status/843172834566332416

**Comment by ara4n** [comment]
might be worth mentioning them in the "what if we did this differently?" section at the end.

**Comment by glyph** [comment]
I wouldn't say that our library is "better" or "less hacky" - it doesn't really address the following-deferreds case either, it's just that this ought to be built in to Twisted :).  After all, you'll want logging from Twisted itself to follow your context, not get lost.

**Comment by glyph** [comment]
More importantly, to really close the loop on this, Twisted will need to change, probably to modify `Deferred`, since functions in Twisted and other libraries will not follow Synapse's Deferred-creation rules :)

File: docs/log_contexts.rst (hunk 1 above)
Comment (richvdh) [inline docs/log_contexts.rst]: ha. No, I didn't. Possibly because the stupid github mobile site doesn't show them :(

**Comment by richvdh** [comment]
Yup, I looked briefly at https://bitbucket.org/hipchat/txlocal when I was wondering why our stuff worked the way it did. I was put off by a few things: having to wrap every single function that gets added as a `callback` sounded fiddly; I was also bit alarmed by glyph's analysis that it dug around deep in twisted internals, and I think I read somewhere that it didn't work on Twisted 16 (and it's not been touched in the last year).

I also came across `twisted.python.context`. Maybe I'm missing the obvious, but I couldn't see how it interacted with deferred - again it looks like you have to wrap everything you `addCallback` (and how it would work with `inlineCallbacks` is even more of a mystery).

But both of those approaches miss the point of this PR. We have a system in Synapse, which works tolerably well, and I don't think abuses the innards of Twisted too badly. If we started again now, we probably wouldn't do it like this - and yes, more support from the Twisted framework would be nice. But given we don't seem likely to be rewriting it any time soon, then for sanity's sake, let's at least have some record of how the current code works.

File: docs/log_contexts.rst (hunk 1 above)
Comment (ara4n) [inline docs/log_contexts.rst]: the github mobile app is *utterly* useless :(

File: docs/log_contexts.rst [hunk 2]
```diff
@@ -1,10 +1,441 @@
...
+deferred, you will want to yield on it. To that end, whereever functions return
+a deferred, we adopt the following conventions:
+
+.. note:: Rules for functions returning deferreds:
+
+ * If the deferred is already complete, the function returns with the same
+   logcontext it started with.
+ * If the deferred is incomplete, the function clears the logcontext before
+   returning; when the deferred completes, it restores the logcontext before
+   running any callbacks.
```
Comment (erikjohnston) [inline docs/log_contexts.rst]: This isn't being rendered.

**Comment by erikjohnston** [comment]
Rendered: https://github.com/matrix-org/synapse/blob/rav/logcontext_docs/docs/log_contexts.rst

File: docs/log_contexts.rst (hunk 2 above)
Comment (erikjohnston) [inline docs/log_contexts.rst]: Err, that's not even true, it just renders funny.

**Comment by erikjohnston** [comment]
lgtm, other than the weird formatting of the `..note` block

**Comment by glyph** [comment]
> But both of those approaches miss the point of this PR.

Sorry I didn't reply sooner -

I was asked for my feedback, and I thought that the request was for feedback on the system as a whole.  I really hope I didn't come across as objecting to documenting code that had already been written!  Just trying to suggest directions for future possible collaboration.  If anything, this documentation will be a very useful resource in doing so.

Thank you for doing this work.

**Comment by richvdh** [comment]
> I really hope I didn't come across as objecting to documenting code that had already been written!

Not at all. Your feedback and ideas are very much appreciated - sorry for not saying so explicitly!

**Pull request merged**